### PR TITLE
cmd/lncli: add closeallchannels command

### LINKED ISSUE #641: [Request] command to close all opened channels
Currently there is no way to close all channels that are opened without some hacking. It would be nice to have something like:

`lncli closechannel -all`

If you want to close all channels: @djseeds:

## to close all channels...  

```lncli listchannels | jq '.[][]' | jq -r '.channel_point' | tr : ' ' | xargs -n2 lncli closechannel```

## to close only inactive channels...

```lncli listchannels | jq '.[][]' | jq -c -r 'select(.active == false) | .channel_point' | tr : ' ' | xargs -n2 lncli closechannel --force```

### PR DESCRIPTION
**Pull request by wilmerpaulino**
Fixes #641.

### CODE REVIEW

**Comment by merehap** [comment]
There's a lot of new code here. Could you add some testing to cover it? Probably unit testing. Of course there's a lot of IO happening so you'd need to mock some things or the equivalent. Thanks!

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -672,51 +679,110 @@ func closeChannel(ctx *cli.Context) error {
...
+			ChannelPoint: &lnrpc.ChannelPoint{
+				FundingTxid: txid[:],
+				OutputIndex: uint32(index),
+			},
+			Force: !channel.Active,
+		}
+
+		closingTxid, err := closeChannelRequest(ctx, req)
+		if err != nil {
+			continue
```
Comment (merehap) [inline cmd/lncli/commands.go]: It would probably be good to do some error reporting if a channel fails to close (especially since in the existing channel closure case, if a channel fails to close, there is an error reported to the user). I'm not sure what the most idiomatic way to do that for this project is, but I'd I see two main ways to do it (and it is possible that both should be done).

1. Print an error message upon each closure here, before the "continue". I'm not sure if printing error messages as you go is idiomatic though.
2. Return an error message at the end of this function if the number of channels that were closed was fewer than the number of channels that were requested to be closed. This might be more idiomatic, but has the down side of the user not knowing why individual channels didn't close.

You'll probably need feedback from a regular contributor on how best to handle this, though I suspect that at least some error reporting is needed.

Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Yeah, some error reporting would definitely be helpful. I wanted to submit this first just for some feedback. I think something like the following response would be helpful:

```
{
    "completed": [
        {
            "remote_pubkey": "...",
            "channel_point": "...",
            "closing_txid": "...",
        },
        {
            "remote_pubkey": "...",
            "channel_point": "...",
            "closing_txid": "...",
        },
    ],
    "failed": [
        {
            "remote_pubkey": "...",
            "channel_point": "...",
            "error": "...",
        },
    ],
}
```

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -672,51 +679,110 @@ func closeChannel(ctx *cli.Context) error {
...
+		}
+		index, err := strconv.ParseUint(s[1], 10, 32)
+		if err != nil {
+			continue
+		}
+
+		req := &lnrpc.CloseChannelRequest{
+			ChannelPoint: &lnrpc.ChannelPoint{
+				FundingTxid: txid[:],
```
Comment (Roasbeef) [inline cmd/lncli/commands.go]: Note that this'll need to be modified to use the `oneof` field once the other PR fixing the issue w.r.t force closing over REST if fixed.

File: cmd/lncli/commands.go (hunk 1 above)
Comment (Roasbeef) [reply cmd/lncli/commands.go]: Indeed. For example, on mainnet we may want to prompt the user if they would like to _really_ force close, showing the number of blocks they'd need to wait until the funds can be swept entirely.

File: cmd/lncli/commands.go [hunk 3]
```diff
@@ -672,51 +679,110 @@ func closeChannel(ctx *cli.Context) error {
...
+	printJSON(struct {
+		ClosingTxid string `json:"closing_txid"`
+	}{
+		ClosingTxid: closingTxid,
+	})
+
+	return nil
+}
+
+func closeChannelRequest(ctx *cli.Context, req *lnrpc.CloseChannelRequest) (string, error) {
```
Comment (Roasbeef) [inline cmd/lncli/commands.go]: Naming suggestion `executeChannelClose`.

**Comment by Roasbeef** [comment]
@merehap much of the underlying code that this command uses is itself tested. Atm, we don't have any infrastructure for testing new _command line utility function_.

**Comment by merehap** [comment]
@Roasbeef OK, then if you are fine with it, I'm going to set up unit tests for all of commands.go. There's a lot of code in commands.go that isn't hit by tests that hit the underlying functionality. I'd like to go with unit tests instead of integration tests since, as you said, the underlying functionality is already tested, and because it is easier to be more thorough with exercising error cases with unit tests. Sound good?

@wilmerpaulino I retract my request for you to add tests. I'll add unit tests after your change is merged in.

**Comment by wilmerpaulino** [comment]
Rebased and addressed comments.

File: cmd/lncli/commands.go [hunk 4]
```diff
@@ -688,53 +695,161 @@ func closeChannel(ctx *cli.Context) error {
...
+			ChannelPoint: &lnrpc.ChannelPoint{
+				FundingTxid: &lnrpc.ChannelPoint_FundingTxidStr{
+					FundingTxidStr: s[0],
+				},
+				OutputIndex: uint32(index),
+			},
+			Force: forceClose,
+		}
+
+		closingTxid, err := executeChannelClose(ctx, req)
```
Comment (halseth) [inline cmd/lncli/commands.go]: I haven't tested this change, but won't this mean that you will have to wait for one `executeChannelClose` to finish before going to the next one? I suspect this can get slow when you have many channels, some of which must be force closed. You also won't see any of the results before all channels finishes closing. 

Possible to ask for confirmation up front, then closing all channels in parallel?

**Comment by wilmerpaulino** [comment]
Thanks for the review @halseth! I've addressed your comments.

File: cmd/lncli/commands.go [hunk 5]
```diff
@@ -707,7 +717,138 @@ func closeChannel(ctx *cli.Context) error {
...
+
+		wg.Done()
+	}()
+
+	if err := executeChannelClose(ctx, req, txidChan); err != nil {
+		return err
+	}
+
+	// We'll wait before exiting to make sure the closing txid is printed.
+	wg.Wait()
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: I think we can accomplish a similar result without using the wait group. Waiting on `txidChan` should block to prevent the goroutine from returning. Why not just wait for the result, print json, then return?

File: cmd/lncli/commands.go [hunk 6]
```diff
@@ -707,7 +717,138 @@ func closeChannel(ctx *cli.Context) error {
...
+
+	return nil
+}
+
+func closeAllChannels(ctx *cli.Context) error {
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: Needs godoc comment

File: cmd/lncli/commands.go [hunk 7]
```diff
@@ -728,28 +869,37 @@ func closeChannel(ctx *cli.Context) error {
...
+		}
+	}
+}
 
-			printJSON(struct {
-				ClosingTXID string `json:"closing_txid"`
-			}{
-				ClosingTXID: txid.String(),
-			})
+func promptForConfirmation(msg string) bool {
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: Godoc comment

File: cmd/lncli/commands.go [hunk 8]
```diff
@@ -707,7 +717,138 @@ func closeChannel(ctx *cli.Context) error {
...
+
+	for i := 0; i < len(channels); i++ {
+		res := <-resultChan
+		printJSON(res)
+	}
+
+	return nil
+}
+
+func executeChannelClose(ctx *cli.Context, req *lnrpc.CloseChannelRequest,
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: Godoc comment

File: cmd/lncli/commands.go [hunk 9]
```diff
@@ -651,27 +656,30 @@ var closeChannelCommand = cli.Command{
...
-
-	// Show command help if no arguments provided
 	if ctx.NArg() == 0 && ctx.NumFlags() == 0 {
 		cli.ShowCommandHelp(ctx, "closechannel")
 		return nil
 	}
 
+	if ctx.Bool("all") {
+		// Make sure only the "all" flag was set.
+		if ctx.NArg() != 0 || ctx.NumFlags() != 1 {
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: Can I suggest that we allow `—all` to be used in conjunction with `—force` so that the user doesn’t have to manually step through each inactive channel? Might even consider adding `—force-only` so that we can keep active channels!

File: cmd/lncli/commands.go [hunk 10]
```diff
@@ -651,27 +656,30 @@ var closeChannelCommand = cli.Command{
...
+			return fmt.Errorf("--all cannot be used with other " +
+				"args/flags")
+		}
+
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: If we could prompt the user here to confirm that they indeed want to close all channels, I think that would go a long way in preventing an apocalypse :)

File: cmd/lncli/commands.go (hunk 5 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Yeah, I gave that a try. Basically, since `executeChannelClose` doesn't block without explicitly setting the `block` flag, we'll exit from `closeChannel` before being able to print the closing txid. I was able to get around that by using the `WaitGroup`, but maybe there's a better way?

File: cmd/lncli/commands.go (hunk 9 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Definitely! I think it'd be better to create another command `closeallchannels` to avoid making the `force` flag have different uses depending on the context. This would also allow us to introduce additional flags, like for closing inactive channels only.

File: cmd/lncli/commands.go (hunk 10 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: If we decide to create a `closeallchannels` command, we could mimic how `rm` works:

`closeallchannels` would ask for confirmation for every inactive channel only and attempt to close all channels.

`closeallchannels --inactive-only` would ask for confirmation for every inactive channel and attempt to close them.

`closeallchannels --force` would only ask for confirmation once and attempt to close all channels.

`closeallchannels --inactive-only --force` would ask for confirmation once and attempt to close all inactive channels only.

File: cmd/lncli/commands.go (hunk 5 above)
Comment (cfromknecht) [reply cmd/lncli/commands.go]: Ahh I see, without the wait group/go routine we wouldn't print the info until after a `block` call has been completed. Ignore my previous comment, I think this is a good workaround!

File: cmd/lncli/commands.go (hunk 10 above)
Comment (cfromknecht) [reply cmd/lncli/commands.go]: Excellent suggestion, I like that our mass channel teardown logic would be done behind a separate rpc altogether. Mimicking `rm` makes sense from a UX perspective, and the flags you suggest work well 😊

Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Is a separate RPC necessary though? The confirmation aspect would be a bit more tricky to handle when using anything other than `lncli`.

**Comment by wilmerpaulino** [comment]
Feedback addressed. This functionality is now implemented as a new command `closeallchannels`.

File: cmd/lncli/commands.go [hunk 11]
```diff
@@ -707,7 +701,41 @@ func closeChannel(ctx *cli.Context) error {
...
+	wg.Add(1)
+	go func() {
+		printJSON(struct {
+			ClosingTxid string `json:"closing_txid"`
+		}{
+			ClosingTxid: <-txidChan,
+		})
+
+		wg.Done()
```
Comment (halseth) [inline cmd/lncli/commands.go]: this can be made a `defer`

File: cmd/lncli/commands.go [hunk 12]
```diff
@@ -680,25 +672,27 @@ func closeChannel(ctx *cli.Context) error {
 		SatPerByte:   ctx.Int64("sat_per_byte"),
 	}
 
+	args := ctx.Args()
+
 	switch {
 	case ctx.IsSet("funding_txid"):
-		txid = ctx.String("funding_txid")
+		req.ChannelPoint.FundingTxid = &lnrpc.ChannelPoint_FundingTxidStr{
```
Comment (halseth) [inline cmd/lncli/commands.go]: is this change needed?

File: cmd/lncli/commands.go [hunk 13]
```diff
@@ -728,28 +756,216 @@ func closeChannel(ctx *cli.Context) error {
...
+		// We can safely exit if the user did not confirm.
+		if !confirmed {
+			return nil
+		}
+
+		for i, channel := range channelsToClose {
+			// If the channel is active, remove it from our list
+			// as we were requested to close inactive channels only.
+			if channel.GetActive() {
+				channelsToClose = append(channelsToClose[:i],
```
Comment (halseth) [inline cmd/lncli/commands.go]: maaaybe it is easier to reason about this code if you start out with an empty slice and then *add* channels to `channelsToClose` instead of deleting them?

File: cmd/lncli/commands.go [hunk 14]
```diff
@@ -728,28 +756,216 @@ func closeChannel(ctx *cli.Context) error {
...
+	// Launch each channel closure in a goroutine in order to execute them
+	// in parallel. Once they're all executed, we will print the results as
+	// they come.
+	resultChan := make(chan result, len(channelsToClose))
+	for _, channel := range channelsToClose {
+		go func(channel *lnrpc.ActiveChannel) {
+			res := result{}
+			res.RemotePubKey = channel.RemotePubkey
+			res.ChannelPoint = channel.ChannelPoint
+			defer func() { resultChan <- res }()
```
Comment (halseth) [inline cmd/lncli/commands.go]: nit: expand this over multiple lines

File: cmd/lncli/commands.go [hunk 15]
```diff
@@ -728,28 +756,216 @@ func closeChannel(ctx *cli.Context) error {
...
-				ClosingTXID: txid.String(),
-			})
+			req := &lnrpc.CloseChannelRequest{
+				ChannelPoint: &lnrpc.ChannelPoint{
+					FundingTxid: &lnrpc.ChannelPoint_FundingTxidStr{
+						FundingTxidStr: s[0],
+					},
+					OutputIndex: uint32(index),
+				},
+				Force: !channel.GetActive(),
```
Comment (halseth) [inline cmd/lncli/commands.go]: You can set `block: false` for readability

File: cmd/lncli/commands.go (hunk 12 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Not really, I just wanted to match how the `output_index` is parsed below.

File: cmd/lncli/commands.go (hunk 11 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Fixed.

File: cmd/lncli/commands.go (hunk 13 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Fixed.

File: cmd/lncli/commands.go (hunk 14 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Fixed.

File: cmd/lncli/commands.go (hunk 15 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: `block` is actually not used in the `lnrpc.CloseChannelRequest`, it's just a flag used by the `closechannel` command.

Comment (halseth) [reply cmd/lncli/commands.go]: Ah 😅

File: cmd/lncli/commands.go [hunk 16]
```diff
@@ -707,7 +701,41 @@ func closeChannel(ctx *cli.Context) error {
 		req.ChannelPoint.OutputIndex = 0
 	}
 
-	stream, err := client.CloseChannel(ctxb, req)
+	txidChan := make(chan string, 1)
```
Comment (halseth) [inline cmd/lncli/commands.go]: The following chunk of code is a bit cryptic to a reader I think. Would be nice to add some comments, making it easier to understand the purpose of the go routine and channels.

File: cmd/lncli/commands.go [hunk 17]
```diff
@@ -707,7 +701,41 @@ func closeChannel(ctx *cli.Context) error {
...
+	wg := sync.WaitGroup{}
```
Comment (halseth) [inline cmd/lncli/commands.go]: nit: troughout the codebase, wait groups are usually declared `var wg sync.WaitGroup`

File: cmd/lncli/commands.go [hunk 18]
```diff
@@ -707,7 +701,41 @@ func closeChannel(ctx *cli.Context) error {
...
+	// In the case that the user did not provide the `block` flag, then we
+	// need to wait for the goroutine to be done to prevent it from being
+	// destroyed when exiting before printing the closing transaction ID.
+	wg.Wait()
+
+	return nil
+}
+
+// executeChannelClose attempts to a close a channel from a request.
+// The closing transaction ID is sent through the txidChan channel.
```
Comment (halseth) [inline cmd/lncli/commands.go]: note that it is sent after the tx is broadcasted (not yet confirmed)

File: cmd/lncli/commands.go [hunk 19]
```diff
@@ -728,28 +756,224 @@ func closeChannel(ctx *cli.Context) error {
...
+					FundingTxid: &lnrpc.ChannelPoint_FundingTxidStr{
+						FundingTxidStr: s[0],
+					},
+					OutputIndex: uint32(index),
+				},
+				Force: !channel.GetActive(),
+			}
+
+			txidChan := make(chan string, 1)
+			err = executeChannelClose(ctx, req, txidChan)
```
Comment (halseth) [inline cmd/lncli/commands.go]: One potential problem with passing in the context here (haven't tested it): what if you call `closeallchannels` with `block=true`? Would maybe be safer to let `executeChannelClose` take the `client` and a `boolean` and in this case just setting it false.

File: cmd/lncli/commands.go [hunk 20]
```diff
@@ -728,28 +756,224 @@ func closeChannel(ctx *cli.Context) error {
...
+				res.FailErr = fmt.Sprintf("unable to close "+
+					"channel: %v", err)
+				return
+			}
+
+			res.ClosingTxid = <-txidChan
+		}(channel)
+	}
+
+	for i := 0; i < len(channelsToClose); i++ {
```
Comment (halseth) [inline cmd/lncli/commands.go]: nit:
```go
for i := range channelsToClose {
```

File: cmd/lncli/commands.go (hunk 19 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Nice catch!

**Comment by wilmerpaulino** [comment]
Rebased.
